lib/active_record_shards/default_slave_patches.rb

```diff
@@ -24,55 +24,55 @@ def #{method}_with_default_slave#{punctuation}(*args, &block)
       RUBY
     end
 
+    def columns_with_default_slave(*args, &block)
+      if on_slave_by_default? && !Thread.current[:_active_record_shards_slave_off]
+        read_columns_from = :slave
+      else
+        read_columns_form = :master
+      end
+
+      on_cx_switch_block(read_columns_from, :construct_ro_scope => false) { columns_without_default_slave(*args, &block) }
+    end
+
+    def transaction_with_slave_off(*args, &block)
+      if on_slave_by_default?
+        old_val = Thread.current[:_active_record_shards_slave_off]
+        Thread.current[:_active_record_shards_slave_off] = true
+      end
+
+      transaction_without_slave_off(*args, &block)
+    ensure
+      if on_slave_by_default?
+        Thread.current[:_active_record_shards_slave_off] = old_val
+      end
+    end
+
+    module InstanceMethods
+      # fix ActiveRecord to do the right thing, and use our aliased quote_value
+      def quote_value(*args, &block)
+        self.class.quote_value(*args, &block)
+      end
+
+      def reload_with_slave_off(*args, &block)
+        self.class.on_master { reload_without_slave_off(*args, &block) }
+      end
+    end
+
     CLASS_SLAVE_METHODS = [ :find_by_sql, :count_by_sql,  :calculate, :find_one, :find_some, :find_every, :quote_value, :sanitize_sql_hash_for_conditions, :exists?, :table_exists? ]
 
     def self.extended(base)
       CLASS_SLAVE_METHODS.each { |m| ActiveRecordShards::DefaultSlavePatches.wrap_method_in_on_slave(true, base, m) }
 
-      base.class_eval do
-        # fix ActiveRecord to do the right thing, and use our aliased quote_value
-        def quote_value(*args, &block)
-          self.class.quote_value(*args, &block)
-        end
-
-        def reload_with_slave_off(*args, &block)
-          self.class.on_master { reload_without_slave_off(*args, &block) }
-        end
-        alias_method :reload_without_slave_off, :reload
-        alias_method :reload, :reload_with_slave_off
-
-        class << self
-          def columns_with_default_slave(*args, &block)
-            if on_slave_by_default? && !Thread.current[:_active_record_shards_slave_off]
-              read_columns_from = :slave
-            else
-              read_columns_form = :master
-            end
-
-            on_cx_switch_block(read_columns_from, :construct_ro_scope => false) { columns_without_default_slave(*args, &block) }
-          end
-          alias_method :columns_without_default_slave, :columns
-          alias_method :columns, :columns_with_default_slave
-        end
+      base.singleton_class.send(:alias_method, :columns_without_default_slave, :columns)
```

Comment:
What I don’t like is this pattern

```rb
module Foo
  def self.extended(base)
    base.class_eval do
      class << self
        def bar
```
much cleaner:
```rb
module Foo
  def bar
```
I’m fine with using `class << self` for those `alias_method` calls.